tests/unit/fem/test_r2d_bilininteg.cpp

```diff
@@ -0,0 +1,4671 @@
+   ddF[0] = dq[1]*dF[2] - dq[2]*dF[1];
+   ddF[1] = dq[2]*dF[0] - dq[0]*dF[2];
+   ddF[2] = dq[0]*dF[1] - dq[1]*dF[0];
+}
+
+void Curl_VcrossGrad_f3(const Vector & x, Vector & ddf)
+{
+   DenseMatrix dV; Grad_V3(x, dV);
+   Vector df;      Grad_f3(x, df);
+   ddf.SetSize(3);
+   ddf[0] = dV(0,1)*df[1] + dV(0,2)*df[2] - (dV(1,1)+dV(2,2))*df[0];
+   ddf[1] = dV(1,2)*df[2] + dV(1,0)*df[0] - (dV(2,2)+dV(0,0))*df[1];
+   ddf[2] = dV(2,0)*df[0] + dV(2,1)*df[1] - (dV(0,0)+dV(1,1))*df[2];
+}
+
+void Curl_VcrossCurlF3(const Vector & x, Vector & ddF)
+{
+   DenseMatrix dv; Grad_V3(x, dv);
+   Vector dF; CurlF3(x, dF);
+   ddF.SetSize(3);
+   ddF[0] = dv(0,1)*dF[1] + dv(0,2)*dF[2] - (dv(1,1)+dv(2,2))*dF[0];
+   ddF[1] = dv(1,2)*dF[2] + dv(1,0)*dF[0] - (dv(2,2)+dv(0,0))*dF[1];
+   ddF[2] = dv(2,0)*dF[0] + dv(2,1)*dF[1] - (dv(0,0)+dv(1,1))*dF[2];
+}
+
+void Curl_DCurlF3(const Vector & x, Vector & ddF)
+{
+   DenseMatrix dv;
+   Grad_V3(x, dv);
+   Vector dF; CurlF3(x, dF);
+   ddF.SetSize(3);
+   ddF[0] = dv(2,1)*dF[2] - dv(1,2)*dF[1];
+   ddF[1] = dv(0,2)*dF[0] - dv(2,0)*dF[2];
+   ddF[2] = dv(1,0)*dF[1] - dv(0,1)*dF[0];
+}
+
+void Curl_MCurlF3(const Vector & x, Vector & ddF)
+{
+   DenseTensor dm;
+   Grad_M3(x, dm);
+   Vector dF; CurlF3(x, dF);
+   ddF.SetSize(3);
+   ddF[0] =
+      (dm(2,0,1)-dm(1,0,2))*dF[0] +
+      (dm(2,1,1)-dm(1,1,2))*dF[1] +
+      (dm(2,2,1)-dm(1,2,2))*dF[2];
+   ddF[1] =
+      (dm(0,0,2)-dm(2,0,0))*dF[0] +
+      (dm(0,1,2)-dm(2,1,0))*dF[1] +
+      (dm(0,2,2)-dm(2,2,0))*dF[2];
+   ddF[2] =
+      (dm(1,0,0)-dm(0,0,1))*dF[0] +
+      (dm(1,1,0)-dm(0,1,1))*dF[1] +
+      (dm(1,2,0)-dm(0,2,1))*dF[2];
+}
+
+void Grad_qDivF3(const Vector & x, Vector & ddF)
+{
+   Grad_q3(x, ddF);
+   ddF *= DivF3(x);
+}
+
+void GradVdotGrad_f3(const Vector & x, Vector & ddf)
+{
+   DenseMatrix dv; Grad_V3(x, dv);
+   Vector df; Grad_f3(x, df);
+   ddf.SetSize(3);
+   dv.MultTranspose(df,ddf);
+}
+
+double DivVDivF3(const Vector & x)
+{
+   return DivV3(x)*DivF3(x);
+}
+
+double DivVcrossCurlF3(const Vector & x)
+{
+   Vector dV; CurlV3(x, dV);
+   Vector dF; CurlF3(x, dF);
+
+   return dV * dF;
+}
+
+TEST_CASE("R2D Bilinear Vector Mass Integrators",
+          "[ND_R2D_FECollection]"
+          "[RT_R2D_FECollection]"
+          "[VectorFEMassIntegrator]"
+          "[MixedVectorMassIntegrator]"
+          "[MixedVectorIntegrator]"
+          "[BilinearFormIntegrator]"
+          "[NonlinearFormIntegrator]")
+{
+   int order = 2, n = 1, dim = 2, vdim = 3;
+   double cg_rtol = 1e-14;
+   double tol = 1e-5;
+
+   VectorFunctionCoefficient  F3_coef(vdim, F3);
+   FunctionCoefficient        q3_coef(q3);
+   VectorFunctionCoefficient  D3_coef(vdim, V3);
+   MatrixFunctionCoefficient  M3_coef(vdim, M3);
+   MatrixFunctionCoefficient MT3_coef(vdim, MT3);
+   VectorFunctionCoefficient qF3_coef(vdim, qF3);
+   VectorFunctionCoefficient DF3_coef(vdim, DF3);
+   VectorFunctionCoefficient MF3_coef(vdim, MF3);
+
+   for (int type = (int)Element::TRIANGLE;
```

Comment:
It looks like the files `test_r1d_bilininteg.cpp` and `test_r2d_bilininteg.cpp` are similar enough that they could be combined into one file, with a loop over all types of elements. Depending on the dimension, a pointer to a `FiniteElementCollection` could be set to a new 1D or 2D collection. Maybe there are complications I don't see, but if this could be done, it would save 4600+ lines of code.

Reply:
I suspected you might comment on this :smirk:

It would probably be possible. One issue would be the tests which mix the new classes with derivatives of the preexisting classes such as gradient of an H1. These tend to involve multiplication with rectangular matrices like 3x1 or 3x2 mapping matrices. The R1D versions tend to have more pieces than the R2D versions which may not be trivial to merge. Another issue is the analytic functions at the top of the page. These are generally named the same in the two files but they have different definitions.  Renaming and merging these would not be fun although it could certainly be done.

The last issue that comes to mind might be the most important for me. In the current implementation the `TEST_CASE` statements can be used to run subsets of the unit tests and this is _so_ convenient when tracking down problems. If we merged these unit tests into a single file we'd need to merge the names appearing in these `TEST_CASE` statements and we would be able to just run R1D or R2D cases separately. This issues has already bit me after we merged other tests and I've had to temporarily edit files so that they skip certain tests so that I can focus on more urgent things. Clearly there are workarounds for this issue and I could manage if the size of the unit test files is something we need to reduce.

In short, I am willing to merge these but I'd rather not.

Reply:
It sounds reasonable to keep them separate then. I think we should merge unless there are good reasons not to, and you provided some good reasons.

Reply:
That's a relief 😌. Thanks, @dylan-copeland !